// homebridge-config-ui-x/dist/bin/base-platform.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.BasePlatform = void 0;
const node_process_1 = __importDefault(require("node:process"));
class BasePlatform {
    constructor(hbService) {
        this.hbService = hbService;
    }
    async install() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async uninstall() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async start() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async stop() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async restart() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async beforeStart() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async rebuild(all = false) {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async viewLogs() {
        this.hbService.logger('This command has not been implemented on this platform.', 'fail');
        node_process_1.default.exit(0);
    }
    async getId() {
        return {
            uid: 0,
            gid: 0,
        };
    }
    getPidOfPort(port) {
        return null;
    }
    async updateNodejs(job) { }
}
exports.BasePlatform = BasePlatform;
